File: dekai_datachannel/_events.py
from typing import Callable


class EventTarget:
    def __init__(self):
        self.__event_handlers: dict[str, list[Callable]] = {}

    def on(self, event: str, handler: Callable):
        if event not in self.__event_handlers:
            self.__event_handlers[event] = []
        self.__event_handlers[event].append(handler)

    def off(self, event: str, handler: Callable):
        if event in self.__event_handlers:
            self.__event_handlers[event].remove(handler)

    def _emit(self, event: str, *args, **kwargs):
        if event in self.__event_handlers:
            for callback in self.__event_handlers[event]:
                callback(*args, **kwargs)

File: ts/events.ts
export type EventHandler<Args extends unknown[]> = (...args: Args) => void;

export class EventEmitter<EventMap extends Record<string, unknown[]>> {
  protected handlers = new Map<keyof EventMap, Set<EventHandler<any>>>();

  on<EventName extends keyof EventMap>(
    event: EventName,
    handler: EventHandler<EventMap[EventName]>,
  ): void {
    let eventHandlers = this.handlers.get(event);
    if (eventHandlers === undefined) {
      eventHandlers = new Set();
      this.handlers.set(event, eventHandlers);
    }
    eventHandlers.add(handler as EventHandler<any>);
  }

  off<EventName extends keyof EventMap>(
    event: EventName,
    handler: EventHandler<EventMap[EventName]>,
  ): void {
    this.handlers.get(event)?.delete(handler as EventHandler<any>);
  }

  protected emit<EventName extends keyof EventMap>(
    event: EventName,
    ...args: EventMap[EventName]
  ): void {
    const eventHandlers = this.handlers.get(event);
    if (eventHandlers === undefined) {
      return;
    }

    for (const handler of eventHandlers) {
      handler(...args);
    }
  }

  protected getHandlers<EventName extends keyof EventMap>(
    event: EventName,
  ): Set<EventHandler<EventMap[EventName]>> | undefined {
    return this.handlers.get(event) as Set<EventHandler<EventMap[EventName]>> | undefined;
  }
}